nemo/collections/common/data/dataset.py

```diff
@@ -286,3 +289,313 @@ def __len__(self):
+        sampling_scales: Optional[Union[float, List[float]]] = None,
+        seed: Optional[int] = None,
+        global_rank: int = 0,
+        world_size: int = 1,
+        pure_random: bool = False,
+        force_monochannel: bool = False,
+        infinity_mode: bool = False,
+        sample_rate: int = 16000,
+        augmentor: Optional['AudioAugmentor'] = None,
+    ):
+        super().__init__()
+
+        self.datasets = datasets
+        self.langs = list(datasets.keys())
+        self.langs_set = set(self.langs)
+        self.lang_iterables = {k: None for k in self.langs}
+        self.lang_kind = {k: None for k in self.langs}
+        self.shuffle = shuffle
+        self.min_duration = min_duration
+        self.max_duration = max_duration
+        self.min_monolingual = min_monolingual
+        self.db_norm = db_norm
+        self.pause_start = pause_start
+        self.pause_join = pause_join
+        self.pause_end = pause_end
+        self.pure_random = pure_random
+        self.force_monochannel = force_monochannel
+        self.infinity_mode = infinity_mode
+        self.global_rank = global_rank
+        self.world_size = world_size
+        self.augmentor = augmentor
+        self.sample_rate = sample_rate
+        self.length = 0
+        if lang_probs is None:
+            self.prob_dict = {l: 1.0 / len(self.langs) for l in self.langs}
+        else:
+            assert len(self.langs) == len(
+                lang_probs
+            ), "Size mismatch between languages and respective probs in CodeSwitchedDataset"
+            self.prob_dict = {l: lang_probs[l] for l in self.langs}
+        self.lang_probs = np.array(list(self.prob_dict.values()))
+        if sampling_scales is not None and not isinstance(sampling_scales, list):
+            self.sampling_scales = {k: sampling_scales for k in self.langs}
+        elif (
+            sampling_scales is not None
+            and isinstance(sampling_scales, list)
+            and len(sampling_scales) == len(self.langs)
+        ):
+            self.sampling_scales = {k: v for k, v in zip(self.langs, sampling_scales)}
+        else:
+            self.sampling_scales = {k: 1 for k in self.langs}
+
+        for lang, dataset in self.datasets.items():
+            isiterable = isinstance(dataset, IterableDataset)
+
+            if isiterable:
+                self.lang_kind[lang] = 'iterable'
+                self.length += int(len(dataset) * self.sampling_scales[lang])
+            else:
+                self.lang_kind[lang] = 'map'
+                self.length += int((len(dataset) // world_size) * self.sampling_scales[lang])
+
+        if seed is not None:
+            np.random.seed(seed)
+
+        # set this to ensure compatibility with models searching for the collate_fn
+        # since this class stores datasets as a dict, not list
+        # self.collate_fn = self.datasets[self.langs[0]].collate_fn
+        if hasattr(self.datasets[self.langs[0]], 'collate_fn'):
```

Comment:
These need a lot of checks - for example don't assume self.datasets is a list in the first place, or that it has 'datasets' again as its argument. Youll need to do multiple hasattr for each line here

Reply:
The constructor specifies that self.datasets is a dict, which is how it's used here. You want me to typecheck that it's actually a dict?

Reply:
Fixed, please check newly committed code